### PR TITLE
Build TensorFlow as monolithic into libdeepspeech.so

### PR DESCRIPTION
**Pull request by lissyx**
Also limit the set of included Ops to avoid wasting size.

### CODE REVIEW

File: native_client/BUILD [hunk 1]
```diff
@@ -45,32 +45,58 @@ genrule(
...
         "//third_party/eigen3",
-        ":deepspeech_utils"
-    ] + if_native_model([":deepspeech_model", "//tensorflow/compiler/tf2xla:xla_compiled_cpu_function"]),
+    ] + if_native_model([
+        "//tensorflow/compiler/tf2xla:xla_compiled_cpu_function",
+        "//tensorflow/compiler/xla/service/cpu:runtime_matmul",
+    ])
+      + if_cuda([
+        "//tensorflow/core:core",
+        "//tensorflow/core/kernels:slice_op_gpu",       # Slice GPU
```
Comment (reuben) [inline native_client/BUILD]: Does `core:core` above include the GPU kernels for all the other OPs needed in the CPU list?

File: native_client/README.md [hunk 2]
```diff
@@ -96,17 +96,16 @@ Bazel defines:
...
+bazel build --config=monolithic -c opt --copt=-O3 --copt=-fvisibility=hidden --define=DS_NATIVE_MODEL=1 --define=DS_MODEL_TIMESTEPS=64 --define=DS_MODEL_FRAMESIZE=494 --define=DS_MODEL_FILE=/tmp/model.ldc93s1.pb //native_client:deepspeech_model //native_client:libdeepspeech.so //native_client:deepspeech_utils //native_client:libctc_decoder_with_kenlm.so //native_client:generate_trie
 ```
 
 Later, when building either `deepspeech` binaries or bindings, you will have to add some extra variables to your `make` command-line (assuming `TFDIR` points to your TensorFlow's git clone):
 ```
-EXTRA_LDFLAGS="-L${TFDIR}/bazel-bin/tensorflow/compiler/xla/ -L${TFDIR}/bazel-bin/tensorflow/compiler/aot/ -L${TFDIR}/bazel-bin/tensorflow/compiler/xla/service/cpu/" EXTRA_LIBS="-ldeepspeech_model -lruntime -lexecutable_run_options -lruntime_matmul"
+EXTRA_LIBS="-ldeepspeech_model"
```
Comment (reuben) [inline native_client/README.md]: Nice :D

File: native_client/BUILD (hunk 1 above)
Comment (lissyx) [reply native_client/BUILD]: `core:core` includes the GPU runtime. I was unable to run on GPU without `slice_op_gpu`, it complained of a missing kernel. Taking a look at higher level of debug with GPU would show a lot of ops executed on the GPU, but if you can check that as well on your side, it's welcome :).

My process was running with `TF_CPP_MIN_LOG_LEVEL` and then using `grep` / `wc` to cound number of occurrences with `GPU` and `CPU` with both runtimes, and compare to the old version. I'll do that again to make sure next monday.

Comment (lissyx) [reply native_client/BUILD]: (tf-venv-master) alexandre@serveur:~/tmp/deepspeech/gpu_onelib$ TF_CPP_MIN_VLOG_LEVEL=1 ./deepspeech /home/alexandre/Documents/codaz/Mozilla/DeepSpeech/DeepSpeech-lissyx/test.frozen.988_e120.LSTM.ldc93s1.pb /home/alexandre/Documents/codaz/Mozilla/DeepSpeech/DeepSpeech-lissyx/data/ldc93s1/LDC93S1_long.wav /home/alexandre/Documents/codaz/Mozilla/DeepSpeech/DeepSpeech-lissyx/data/alphabet.txt /home/alexandre/Documents/codaz/Mozilla/DeepSpeech/DeepSpeech-lissyx/data/lm/lm.binary /home/alexandre/Documents/codaz/Mozilla/DeepSpeech/DeepSpeech-lissyx/data/lm/trie -t 2>tf_vlog.txt
she had your dad your dad your dad your dad your dad your dark suit in greasy wash water all year
cpu_time_overall=1.50706 cpu_time_mfcc=0.00629 cpu_time_infer=1.50077
(tf-venv-master) alexandre@serveur:~/tmp/deepspeech/gpu_onelib$ cd ../gpu_legacy/
(tf-venv-master) alexandre@serveur:~/tmp/deepspeech/gpu_legacy$ TF_CPP_MIN_VLOG_LEVEL=1 ./deepspeech /home/alexandre/Documents/codaz/Mozilla/DeepSpeech/DeepSpeech-lissyx/test.frozen.988_e120.LSTM.ldc93s1.pb /home/alexandre/Documents/codaz/Mozilla/DeepSpeech/DeepSpeech-lissyx/data/ldc93s1/LDC93S1_long.wav /home/alexandre/Documents/codaz/Mozilla/DeepSpeech/DeepSpeech-lissyx/data/alphabet.txt /home/alexandre/Documents/codaz/Mozilla/DeepSpeech/DeepSpeech-lissyx/data/lm/lm.binary /home/alexandre/Documents/codaz/Mozilla/DeepSpeech/DeepSpeech-lissyx/data/lm/trie -t 2>tf_vlog.txt
she had your dad your dad your dad your dad your dad your dark suit in greasy wash water all year
cpu_time_overall=1.65676 cpu_time_mfcc=0.00650 cpu_time_infer=1.65027

Comment (lissyx) [reply native_client/BUILD]: And then, comparing runs:
```
(tf-venv-master) alexandre@serveur:~/tmp/deepspeech$ grep -ic "device:GPU:0" gpu_legacy/tf_vlog.txt 
18077
(tf-venv-master) alexandre@serveur:~/tmp/deepspeech$ grep -ic "device:CPU:0" gpu_legacy/tf_vlog.txt 
16
(tf-venv-master) alexandre@serveur:~/tmp/deepspeech$ ls -hal gpu_legacy/libdeepspeech.so
-r-xr-xr-x 1 alexandre alexandre 788K janv.  5 11:50 gpu_legacy/libdeepspeech.so
(tf-venv-master) alexandre@serveur:~/tmp/deepspeech$ grep -ic "device:GPU:0" gpu_onelib/tf_vlog.txt 
18077
(tf-venv-master) alexandre@serveur:~/tmp/deepspeech$ grep -ic "device:CPU:0" gpu_onelib/tf_vlog.txt 
16
(tf-venv-master) alexandre@serveur:~/tmp/deepspeech$ ls -hal gpu_onelib/libdeepspeech.so
-r-xr-xr-x 1 alexandre alexandre 315M janv.  4 15:09 gpu_onelib/libdeepspeech.so
```

**Comment by reuben** [comment]
Do we want to keep the bazel patch and related branches (is_patched_bazel) in the tree? I don't have a strong opinion either way, but it seems like it doesn't have to be on the tree to be useful.

**Comment by lissyx** [comment]
@reuben That's one question I wanted to discuss about. I agree the patch itself might be too much, but the overall idea was to make it easy to switch back to heavy debug mode in case of regression. So I added the patch to make sure we keep it. I could just fork bazel repo and apply the patch there, and upload linux binary, if you prefer. But the `is_patched_bazel` would still be useful IMHO, saves quite some time in debugging.

**Comment by reuben** [comment]
That sounds like a good enough reason to keep it.

**Pull request merged**